v1/v80.c

```diff
@@ -0,0 +1,666 @@
+/* v80.c (C) Gary V. Vaughan 2024, MIT License
+
+   On Linux, compile with:
+   $ cc -o bin/v80 v1/v80.c
+
+   Other architectures pending - from memory so far, this should all be C89 compatible.
+*/
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#ifndef NO_SYS_PARAM_H
+#  include <sys/param.h>
+#endif
+
+#ifndef EXIT_SUCCESS
+#  define EXIT_SUCCESS 0
+#  define EXIT_FAILURE 1
+#endif
+#ifndef MAXPATHLEN
+#  define MAXPATHLEN 1024
+#endif
+
+#define NOTHING_ELSE
+#define NUMBER_MAX          0xffff
+#define TOKEN_MAXLEN        31
+
+
+
+
+/* TYPEDEFS */
+
+
+/* C code sometimes uses `int` as a boolean, and we want to be able to
+   `return (x == y)` from a boolean function without casting or using
+   the ternary operator, so we use these macros to document those APIs. */
+#ifdef Bool
+#  undef Bool
+#endif
+#ifndef TRUE
+#  define TRUE -1
+#endif
+#ifndef FALSE
+#  define FALSE 0
+#endif
+
+typedef int Bool;
+
+typedef struct includestack {
+    struct includestack *next;
+    FILE *stream;
+    const char *zfname;
+    unsigned line, col, indent;
+    Bool isindenting;
+} Include;
+
+typedef struct symbolchain {
+    struct symbolchain *next;
+    char *zname;
+    uint16_t value;
+} Symbol;
+
+typedef struct {
+    const char *zname;
+    const char *zfname;
+    unsigned line, col, indent;
+} Token;
+
+/* TODO: write some functions so we can cast Include and Symbol chains to this
+         type for generic operations. */
+typedef struct stack {
+    struct stack *next;
+} Stack;
+
+
+
+
+/* GLOBALS */
+
+
+Symbol     *dollar  = NULL; /* fast access to the '$' symbol */
```

Comment:
This is known as the "virtual program-counter" and should probably be called `pc`. It's the position in memory the assembled binary thinks it is when it is running. I've tried to use consistent terminology in the v80 source code, readme and other files.

Reply:
Interesting!  To me, PC is the live program counter in the CPU pointing to the address from which the next instruction will be read... I've usually called the address where the next instruction will be written `dollar` since that's the symbol most assemblers use;  in Forth, it's called `HERE` and is the location that various `,` adjacent operations write data before incrementing.

Happy to rename it to PC for consistency if you'd prefer, but I think that's a mildly confusing terminology, personally.

Reply:
Just to double check that I understand that properly: `$` is an offset into the code-segment, so that as we output assembled bytes they accumulate into the code-segment, which is finally flushed to the output file after all the input is exhausted.